src/thunks.js

```diff
@@ -39,36 +39,42 @@ export function makeWSConnection() {
   return (dispatch) => {
     const authToken = getAuthToken();
     const url = 'ws://' + window.location.host + '/api/timers/ws-connect/';
-    let wsConnection
-      = new WebSocket(url);
+    let intervalId;
+    let wsConnection = new WebSocket(url);
+
     wsConnection.onopen = (e) => {
       wsConnection.send(JSON.stringify({
         command: 'authenticate',
         token: authToken
       }));
     };
+
     wsConnection.onerror = (e) => {
       dispatch(wsConnectionFailed());
     };
+
     wsConnection.onmessage = (e) => {
       let message = JSON.parse(e.data);
       if (message['auth-status'] === 'success') {
-        const intervalId = window.setInterval(() => sendPing(wsConnection),
-                                              10000);
+        dispatch(wsReceivedMessage(JSON.parse(e.data)));
+        dispatch(wsConnectionReady(wsConnection));
+        // Begin ping-pong at 10s intervals.
+        intervalId = window.setInterval(() => sendPing(wsConnection), 10000);
+        // Redefine onmessage to stop checking authentication status.
         wsConnection.onmessage = (e) => {
           dispatch(wsReceivedMessage(JSON.parse(e.data)));
-        };
-
-        wsConnection.onclose = (e) => {
-          window.clearInterval(intervalId);
-        };
-
-        dispatch(wsConnectionReady(wsConnection));
+        }
       }
       else {
         dispatch(wsHandshakeFailed());
       }
-    }
+    };
+
+    wsConnection.onclose = (e) => {
+      window.clearInterval(intervalId);
+      // Try reconnecting after 3s.
+      setTimeout(() => dispatch(makeWSConnection(authToken)), 3000)
```

Comment:
And, _makeWSConnection_ can cause the frontend to blow up if put in the wrong place due to a _callback hell_. It can also blow up the stack if setTimeout isn't used: without setTimeout, it holds on to the calling scope to maintain closure, making tons of requests and the WebSocket objects for all of them remains in the browser memory. One of those places where JS is being JS. tbh, I find it hard to write these things with confidence.